flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -1200,7 +1202,7 @@ public void grantLeadership(final UUID newLeaderSessionID) {
 
             startServicesOnLeadership();
 
-            return prepareLeadershipAsync().thenApply(ignored -> true);
+            return prepareLeadershipAsync().thenApply(ignored -> hasLeadership = true);
```

Comment:
Hmm. I think we could set `hasLeadership` to `true` after the `ResourceManager` confirms the leadership. Right?

Reply:
I'm setting `hasLeadership` to `true` here because this is the last step in the rpc main thread. Confirming leadership happens on the io thread, and should trigger fatal error shutdown once failed.